File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        ArrayList<String> cache = new ArrayList<>();
        
        if (cacheSize == 0) {
            return cities.length * 5;
        }
        
        for (int i = 0; i < cities.length; i++) {
            String city = cities[i].toUpperCase(); // 대소문자 구분 안하기 위해서
            if (cache.contains(city)) { // 캐시에 city가 있다면
                cache.remove(city); // city 지우고
                cache.add(city); // 맨 뒤에 다시 추가
                answer += 1;
            }
            
            else { // 캐시에 city가 없다면
                if (cache.size() == cacheSize) { // 캐시 꽉차있으면
                    cache.remove(0); // 맨 처음 들어온(가장 적게 사용된) 도시 지우기
                    cache.add(city);
                    answer += 5;
                }
                else { // 캐시가 여유 있으면
                    cache.add(city);
                    answer += 5;
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
from collections import deque
def solution(cacheSize, cities):
    answer = 0
    if cacheSize == 0:
        answer = 5 * len(cities)
        return answer
    
    q = deque()
    for city in cities:
        city = city.lower()
        # city가 q에 포함되어 있는 경우
        if city in q:
            answer += 1
            q.remove(city)
            q.append(city)
        # city가 q에 포함되어 있지 않은 경우
        else:
            # 캐시가 다 찬 경우
            if len(q) == cacheSize:
                q.popleft()
                q.append(city)
                answer += 5
            else:
                q.append(city)
                answer += 5
        
    return answer